### PR TITLE
[Diffusion] [AMD] Online MXFP4 and FP8 Quantization for Multimodal Generation

### PR DESCRIPTION
**Pull request by ColinZ22**
## Motivation

Adding Online MXFP4 (For AMD GPUs) and FP8 Quantization for multimodal (image and video) generation with models like Z-Image-Turbo and Wan 2.2.

## Modifications

- New `--quantization` server argument allowing loading unquantized model and quantizing weights and activations to MXFP4.
- New `--quantization-ignored-layers` server argument allows skipping certain layers for online quantization (keeping in full precision)
- New `Mxfp4Config` and `Mxfp4LinearMethod` classes utilizing AITER dynamic MXFP4 quantization and MXFP4 GEMM kernels.
- Enabling FP8 online quantization via `--quantization`.

## Usage Example
To online quantize a Diffusion Model to FP8 or MXFP4, simply add the `--quantization` argument: 

```
sglang generate \
  --model-path Tongyi-MAI/Z-Image-Turbo \
  --prompt "A beautiful sunset over the mountains" \
  --save-output
  --quantization fp8
```
```
sglang generate \
  --model-path Tongyi-MAI/Z-Image-Turbo \
  --prompt "A beautiful sunset over the mountains" \
  --save-output
  --quantization mxfp4
```

## Prompt 1: "A cat sitting at the top of a mountain looking down at a futuristic city"
| FP16 | FP8 | MXFP4 |
|------|-----|-------|
| FP16 | FP8 | MXFP4 |
---
## Prompt 2: "A crowd of people of various age at a busy outdoor marketplace"
| FP16 | FP8 | MXFP4 |
|------|-----|-------|
| FP16 | FP8 | MXFP4 |
---
## Prompt 3: "A young child blowing dandelion seeds, golden hour lighting"
| FP16 | FP8 | MXFP4 |
|------|-----|-------|
| FP16 | FP8 | MXFP4 |
---
## Prompt 4: "A city street at sunset with snow-capped mountain in the distant background"
| FP16 | FP8 | MXFP4 |
|------|-----|-------|
| FP16 | FP8 | MXFP4 |
---

## Performance Benchmarking

**Model:** [Z-Image-Turbo](https://huggingface.co/Tongyi-MAI/Z-Image-Turbo)
**Dataset:** 200 images from [HuggingFace Parti-Prompts](https://huggingface.co/datasets/nateraw/parti-prompts)

| Online Quant Method | Transformer Size (GB) | Peak Mem Size (GB) | Total Gen Time (sec) | Denoise Time (sec) | Avg CLIP Score (↑) |
|--------|----------------------|--------------------|----------------------|--------------------|--------------------|
| bf16 (baseline)   | 11.46                | 19.00              | 201.06               | 132.54             | 32.20              |
| fp8    | 5.86 (-49%)          | 13.42 (-29%)       | 191.91 (-5%)         | 131.74 (-1%)       | 32.31              |
| mxfp4  | 3.23 (-72%)          | 10.77 (-43%)       | 165.05 (-18%)        | 104.78 (-21%)      | 32.22              |

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/loader/fsdp_load.py [hunk 1]
```diff
@@ -408,6 +408,8 @@ def load_model_from_full_model_state_dict(
         "bias",
         "norm_q",
         "norm_k",
+        "weight_scale",  # FP8/MXFP4 quantization scales
+        "input_scale",   # FP8/MXFP4 activation quantization scales
```
Comment (BowenBao) [inline python/sglang/multimodal_gen/runtime/loader/fsdp_load.py]: dup of L407

File: python/sglang/multimodal_gen/runtime/server_args.py [hunk 2]
```diff
@@ -757,6 +761,18 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
...
 
+        parser.add_argument(
+            "--quantization",
+            type=str,
+            default=ServerArgs.quantization,
+            help=(
+                "Quantization method for online quantization. "
+                "Options: 'fp8', 'mxfp4'. "
+                "Apply online weight and activation quantization to unquantized models. "
+                "Note: MXFP4 requires ROCm and MI300X (gfx95x)."
```
Comment (BowenBao) [inline python/sglang/multimodal_gen/runtime/server_args.py]:
```suggestion
                "Note: MXFP4 requires ROCm and MI350+ (gfx95x)."
```

File: python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    def apply(
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        bias: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+
+        if not mxfp_supported():
+            raise RuntimeError(
+                "MXFP4 inference requires ROCm MI300X+ (gfx95x). "
```
Comment (BowenBao) [inline python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py]: ditto

File: python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+from sglang.multimodal_gen.runtime.models.parameter import (
+    ModelWeightParameter,
+    PerTensorScaleParameter,
+)
+from sglang.srt.utils import is_hip, mxfp_supported
+
+logger = logging.getLogger(__name__)
+_is_hip = is_hip()
+
+if _is_hip:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py:24]: could we avoid that?

File: python/sglang/multimodal_gen/runtime/server_args.py [hunk 5]
```diff
@@ -757,6 +761,18 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
             help="Disable autocast for denoising loop and vae decoding in pipeline sampling",
         )
 
+        parser.add_argument(
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/server_args.py:1150]:
```
      "Quantization method for the transformer. If omitted, the method is "
      "auto-detected from the checkpoint config or safetensors metadata when "
      "possible. Applies to both pre-quantized checkpoints and online "
      "quantization. Use this flag to override auto-detection. "
      "Options: 'fp8', 'mxfp4'. "
      "Nunchaku SVDQuant is configured separately via --enable-svdquant and "
      "related flags. "
      "Note: MXFP4 requires ROCm and MI350+ (gfx95x)."
```

**Review by mickqian — COMMENTED** [review]
could you also mention this new server arg in `cli.md`, `quantization.md` or other related places?

File: python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py (hunk 4 above)
Comment (ColinZ22) [reply python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py:24]: Can you please elaborate on what you are referring to? Thanks!

**Comment by ColinZ22** [comment]
> could you also mention this new server arg in `cli.md`, `quantization.md` or other related places?

Added documentation in `cli.md` and `quantization.md`

File: python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py (hunk 4 above)
Comment (ColinZ22) [reply python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py:24]: @mickqian

Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py:24]: avoid these locals

Comment (ColinZ22) [reply python/sglang/multimodal_gen/runtime/layers/quantization/mxfp4.py:24]: Hi @mickqian , please correct me if I have misunderstood, but it seems that this pattern of caching `is_hip()` into a module-level `_is_hip` local is a convention across the codebase of sglang (e.g. [activation.py](https://github.com/sgl-project/sglang/blob/0882f5c132fac77e9f55bfea60d4d223e34efc66/python/sglang/multimodal_gen/runtime/layers/activation.py#L17) and [fp8_utils.py](https://github.com/sgl-project/sglang/blob/e8c6e5466c762bc8506bd3a2f799532746d68cb3/python/sglang/srt/layers/quantization/fp8_utils.py#L53)). 
The purpose here is the same: we cache the result so we can use `_is_hip` to conditionally import the aiter MXFP4 kernels (which are only available on ROCm/AMD GPUs). This avoids calling `is_hip()` repeatedly.

**Comment by ColinZ22** [comment]
@mickqian Friendly ping for review, all comments addressed, hoping to land this PR soon!

**Review by BowenBao — APPROVED** [review]
LGTM.

@mickqian , @HaiShaw , @avjves This PR covers a superset of the functionality in #23373. Would it make sense to consolidate the effort and land this one instead? We’re happy to rebase if #23373 lands first, though it seems unnecessary from our perspective.

**Comment by avjves** [comment]
> LGTM.
> 
> @mickqian , @HaiShaw , @avjves This PR covers a superset of the functionality in #23373. Would it make sense to consolidate the effort and land this one instead? We’re happy to rebase if #23373 lands first, though it seems unnecessary from our perspective.

Definitely, I'm happy either way as long as the functionality lands! I originally didn't notice this PR before I had already created a new one.

File: docs/diffusion/quantization.md [hunk 6]
```diff
@@ -40,14 +41,67 @@ Here, `quant_family` means a checkpoint and loading family with shared CLI
...
 | `nunchaku-svdq`   | Pre-quantized Nunchaku transformer weights, usually named `svdq-{int4\|fp4}_r{rank}-...`   | `--transformer-weights-path`                                           | Model-specific support such as Qwen-Image, FLUX, and Z-Image | `nunchaku`                            | SGLang can infer precision and rank from the filename and supports both `int4` and `nvfp4`                                             |
 | `msmodelslim`     | Pre-quantized msmodelslim transformer weights                                              | `--model-path`                                                         | Wan2.2 family                           | None                                  | Currently only compatible with the Ascend NPU family and supports both `w8a8` and `w4a4`                                               |
 
+## Online Quantization
+
+Online quantization applies quantization to unquantized models at load time. This is useful for when pre-quantized checkpoints are not available.
```
Comment (mickqian) [inline docs/diffusion/quantization.md:56]: nit: add (on-the-fly / load-time quantization) as well

**Review by mickqian — APPROVED** [review]

**Comment by avjves** [comment]
@ColinZ22 PR #20922 was merged, which adds the initial support for online quantization, including FP8 quantization. It's missing MXFP4 quantization still though. Are you planning on updating this PR to match the current state to get it merged? :)

**Comment by ColinZ22** [comment]
> @ColinZ22 PR #20922 was merged, which adds the initial support for online quantization, including FP8 quantization. It's missing MXFP4 quantization still though. Are you planning on updating this PR to match the current state to get it merged? :)

@avjves Updated, thanks for letting me know!

**Comment by BowenBao** [comment]
@ColinZ22 please fix lint checks

**Comment by ColinZ22** [comment]
Fixed, @mickqian @wisclmy0611 Re-review would be greatly appreciated! Hoping to land this PR soon.

**Comment by BowenBao** [comment]
@amd-bot ci-status

**Comment by ColinZ22** [comment]
@amd-bot ci-status

**Comment by HaiShaw** [comment]
@ColinZ22 FP8 path is broken on main - FYI.
cc @yichiche @yctseng0211

**Comment by yichiche** [comment]
@HaiShaw Here is the PR to re-activate fp8 aiter backend. https://github.com/sgl-project/sglang/pull/26261

**Comment by yichiche** [comment]
@ColinZ22 Currently we will see error if we --enable-torch-compile true with --quantization fp8, any suggestion?

**Comment by ColinZ22** [comment]
Hi @yichiche, could you try adding `@torch.compiler.disable` to `Fp8LinearMethod.apply()`? This creates a graph break to avoid torch compile crash due to Inductor can't lower `aten._scaled_mm` used in the FP8 GEMM op. 

It should fix the error and allow torch compile to bring significant performance improvements. (For the online MXFP4 quantization path [another fix](https://github.com/ROCm/aiter/pull/2781) is needed to enable torch compile; SGLang Diffusion disables torch compile for all paths by default due to issues like this)

**Comment by avjves** [comment]
One more thing regarding this; am I missing something or is the `quantization-ignored-layers` cli parameter ignored entirely? The value goes to `quantization_ignored_layer`, but nothing seems to be using it?
